### ISSUE TITLE
Add support for `q8` during neural network inference

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#8]
Based on some recent benchmarks (on GTX 1080 Ti), using 8-bit quantized integers during inference could give us significant performance improvement. No experiments has been performed yet, but other sources suggest that the drop in accuracy from using `q8` instead of `f16` (or `f32`) is insignificant.

This is inline with what has been cited by Google and NVIDIA, who are pushing `q8` during inference in their [TPU](https://en.wikipedia.org/wiki/Tensor_processing_unit)'s and [TensorRT](https://developer.nvidia.com/tensorrt) library.

**Updated with new benchmarks using a fused conv-add-act kernel**

## `BATCH_SIZE=1`

```
test f32_128_winograd             ... bench:      23,768 ns/iter (+/- 26)
test f32_256_winograd             ... bench:      75,374 ns/iter (+/- 64)
test i8_128_implicitprecompgemm   ... bench:      32,795 ns/iter (+/- 60)
test i8_256_implicitprecompgemm   ... bench:      54,868 ns/iter (+/- 57)
test i8x4_128_implicitprecompgemm ... bench:      34,073 ns/iter (+/- 58)
test i8x4_256_implicitprecompgemm ... bench:      56,966 ns/iter (+/- 67)
```

Speed-up is 0.70x

## `BATCH_SIZE=16`

```
test f32_128_winograd             ... bench:     163,904 ns/iter (+/- 411)
test f32_256_winograd             ... bench:     557,587 ns/iter (+/- 3,755)
test i8_128_implicitprecompgemm   ... bench:      53,362 ns/iter (+/- 155)
test i8_256_implicitprecompgemm   ... bench:     187,492 ns/iter (+/- 179)
test i8x4_128_implicitprecompgemm ... bench:      54,799 ns/iter (+/- 97)
test i8x4_256_implicitprecompgemm ... bench:     190,671 ns/iter (+/- 137)
```

Speed-up is 3.07x

## `BATCH_SIZE=64`

```
test f32_128_winograd             ... bench:     600,183 ns/iter (+/- 686)
test f32_256_winograd             ... bench:   2,153,691 ns/iter (+/- 2,246)
test i8_128_implicitprecompgemm   ... bench:     175,793 ns/iter (+/- 94)
test i8_256_implicitprecompgemm   ... bench:     609,676 ns/iter (+/- 6,896)
test i8x4_128_implicitprecompgemm ... bench:     178,419 ns/iter (+/- 117)
test i8x4_256_implicitprecompgemm ... bench:     616,550 ns/iter (+/- 12,348)
```

Speed-up is 3.41x

## `BATCH_SIZE=128`

```
test f32_128_winograd             ... bench:   1,180,420 ns/iter (+/- 2,734)
test f32_256_winograd             ... bench:   4,310,177 ns/iter (+/- 4,477)
test i8_128_implicitprecompgemm   ... bench:     326,474 ns/iter (+/- 5,199)
test i8_256_implicitprecompgemm   ... bench:   1,164,039 ns/iter (+/- 3,448)
test i8x4_128_implicitprecompgemm ... bench:     342,554 ns/iter (+/- 1,579)
test i8x4_256_implicitprecompgemm ... bench:   1,184,267 ns/iter (+/- 360)
```

Speed-up is 3.62x

**Comment by Chicoryn** [comment]
Looking at the [compute capabilities table](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions) I suggest the following code paths:

* **7.0**, **6.2** and **5.3** use `f16`
* Everything else use `q8`

**Comment by Chicoryn** [comment]
Some questions remain in how to implement this:

* How do we scale `q8`'s initially and between layers, or do we use the same scale from the start to the end?
* How do we handle the feed-forward layers at the end (matrix multiplication)?
* How do we handle the batch normalization? Do we need to cast the tensor to `f32`, or can we cast the weights? (For `f16` we had to keep the weights as `f32`)

Some articles that may have suggestions:

[Fixed Point Quantization of Deep Convolutional Networks](https://arxiv.org/pdf/1511.06393.pdf)

**Comment by Chicoryn** [comment]
Looking at the trained weights, means, and variance on TensorBoard suggests the following:

* The **input** only have elements with a value of 0.0 or 1.0.
* The **weights** are typically in the range [-0.2, +0.2], with a lower bound of [-0.01, +0.01] and an upper bound of [+0.3, -0.3].
* The **mean** in batch normalization layers is between [-5,+5] with most of them being centered on zero.
* The **variance** in batch normalization layers is all over the place, many are within [0,4] but there are some at [0,128].

On average this would suggest that a range of [-6,+6] is fine (this would give us a step size of 0.047), but there are a few notable exceptions:

1. The _policy head_ downsampling layer would get destroyed since it is squishing 128 features into two features, resulting in fairly large values in the output.
2. The _value head_ downsample layer would get destroyed since it is squishing 128 features into one feature, resulting in fairly large values in the output.
3. Some of the residual layer exhibit near zero and very large weights, these would effectively get pruned by such a range.

Not sure how to handle this since even if we use a dynamic range, if we have to use the same range for the input and the output of a layer then one of them will get pruned if the input and output requires very different ranges.

**Comment by Chicoryn** [comment]
After doing some profiling with [nvprof](http://docs.nvidia.com/cuda/profiler-users-guide/index.html) we can observe that performing the convolution takes up about 99% of the CUDA runtime. This means that we could just convert the tensors to `f32` at the last convolution step and perform the rest in single precision.

Alternatively we need to find some way to convert the `q32` output from `cublasGemmEx` back into `q8` in some efficient way since there is no way to get `q8` output from `cublasGemmEx`.

**Comment by Chicoryn** [comment]
Complications has arisen, from the cuDNN documentation for using the `q8` data type in cross correlations:

```‣ Input and output features maps must be multiple of 4```

This is an issue because our input tensor has 34 features, which is not a multiple of 4. We could remove, or add, two features to fulfill this criteria. There are padding solutions but they feel slow since we need to pad each element in the batch.

Alternatively we could just not do the first layer with `q8`, and do it in `f32` instead, but then we will have issues converting from `f32` to `q8` on the GPU.

There are also issues with the tensor format, since we store everything in `NCHW` because it is the recommended (and fastest) format for `f32` and `f16` code. However for `q8` we need to store everything in `NHWC`, this is not too much of a problem for the input features, but for the weights this is a problem because we do not know their dimensions in the loader.

**Comment by Chicoryn** [comment]
A mock-up implementation gives the following benchmark results, I had to rip out some of the features to accomplish it and decided on the two territory features since I've long disliked them for _giving the network the wrong idea_ anyway.

```
test batch_size_01 ... bench:   1,635,884 ns/iter (+/- 8,217)
test batch_size_02 ... bench:   1,660,229 ns/iter (+/- 10,207)
test batch_size_04 ... bench:   1,673,208 ns/iter (+/- 11,483)
test batch_size_08 ... bench:   1,708,661 ns/iter (+/- 17,706)
test batch_size_16 ... bench:   2,522,723 ns/iter (+/- 10,377)
test batch_size_32 ... bench:   4,780,344 ns/iter (+/- 17,684)
test batch_size_64 ... bench:   8,372,624 ns/iter (+/- 319,169)
```

Speed-up is between 2.27x to 3.20x. This is inline with what I expected from the initially posted benchmarks.

### Implementation notes

* The JSON weight file format now includes a _shape_ field, this is necessary because we need to convert the weights to _NHWC_ format when using `q8`.
* Two features were dropped from the input tensor, so that we now have a multiple of 4.
* Only the intermediate tower and the down-sampling layers are done in `q8`, the rest of the value and policy head and in `f32`. Because of GEMM data type complications.
* This implementation uses the scale [-1,+1] for the value tensor, but uses individual scales for each weight tensor (based on their standard deviation).

**Comment by Chicoryn** [comment]
After doing some sensitivity analysis on using quantized data types it has become pretty clear that in order to not have an explosion of errors we must scale both the input and output correctly in addition to the weights.

This is annoying because it means we must store the appropriate scales for **each** layer. In addition, the scales will change every time we train a new network, making it annoying to support plug-and-play. This leaves us with us with two options:

* Ignore this, since we use batch normalization anyway and this should ensure the output from convolutional layers are _usually_ in the range [-2,+2].
* Compute the appropriate scales at run-time.
* Embed the appropriate scales into the weight file, and keep track of activations (using a moving average) during training.

## Sensitivity analysis of `quantized_types.rs`

As can be seen in Table 1, the maximum absolute error differs by as much as 6x just depending on some minimal changes in the input and output scale. Where the final entry in the table represent the _optimal_ scales.

|`FILTER_SCALE`	|`INPUT_SCALE`	|`OUTPUT_SCALE`	|Absolute error
|---------------|---------------|---------------|--------------
|0.1		|1.0		|1.3		|0.011574924
|0.1		|0.5		|1.3		|0.011574924
|0.1		|0.5		|3.0		|0.012755904
|0.1		|3.0		|3.0		|-
|0.1		|2.0		|3.0		|0.030472517
|0.1		|1.0		|3.0		|0.020314991
|0.1		|0.5		|2.0		|0.011259839
|0.09		|0.4		|2.0		|0.007244095
|0.09		|0.4		|0.81		|0.0063780546

**Comment by Chicoryn** [comment]
Some unofficial documentation on how `cudnnConvolutionBiasActivationForward` works with quantized types. It computes the following formula where `C` is the convolution operator and `act` is the activation operator:

```
y = act(α · C(W, x) + β · z + b)
```

When computing using quantized inputs and outputs the variables will have the following types and ranges:

* `α ∈ f32`
* `β ∈ f32`
* `W ∈ [-127,+127] ∈ i8`
* `x ∈ [-127,+127] ∈ i8`
* `y ∈ [-127,+127] ∈ i8`
* `b ∈ [-127,+127] ∈ f32`

For the most part this is not very surprising, with the exception of _b_, which is scaled to the quantized range as an `f32`. In addition the function performs **no** automatic re-scaling of any of the values, all re-scaling has to be done using α and β.

To determine what α and β should be, one can substitute in the scaling from `f32` to `i8` into the formula where `s?` is a scaling factor in the range `(0,Inf]` and solve for the output `127 y / sy`:

```
act(α · C(127 W / sW, 127 x / sx) + β · 127 z / sz + 127 b / sb)
= act(α · 127² C(W / sW, x / sx) + β · 127 z / sz + 127 b / sb)
= 127 act(α · 127 C(W / sW, x / sx) + β · z / sz + b / sb)
```

After some simplification (above) we can see the 127 terms on both sides cancel each other out, but we still need to sort out the scaling factors. Since `act` is linear we can break this into three sub-problems where each problem must end up with the unit `1 / sy`.

* `α · 127 C(W / sW, x / sx)` implies that `α = (sW sx) / (127 sy)`.
* `β · z / sz` implies that `β = sz / sy`.
* `b / sb` implies that `sb = sy`.

If the scaling factors for `x`, `y`, and `z` are all the same (as one would expect if every layer is batch normalized). Then this simplifies to, where `s` is the _universal_ scaling factor:

* `α = sW / 127`.
* `β = 1`.
* `sb = s`.

**Comment by Chicoryn** [comment]
Been experimenting with computing the scale using a static analysis of the mean and variance of the input tensors since if make the following assumptions about the convolution operator function:

```

**Comment by Chicoryn** [comment]
The introduction of convolutions with dilation to each residual block has showed great success when it comes to global thinking. Unfortunately there is no support for `dilation > 1` when using the data type `q8`. So if we want to use convolutions with dilation then we cannot use quantized types.

The main workaround for this would be to write our own custom kernels that operate on `q8`, this would be good for several other reasons:

* Reduce the overhead associated with kernel launches as we can write a larger fused kernel.
* The convolution kernel for dilation is using the `ImplicitPrecompGemm` algorithm instead of `Winograd`, it should be possible to write an implementation of Winograd for dilation, but presumably NVIDIA just did not bother.

## Clipped rectified linear units

I have been playing around with `tf.nn.relu6` instead of `tf.nn.relu` in the neural network to side-step the _scale problem_ that we have been trying to solve above. So far it has proven very successful as the change in loss is within margin for error, but it solves the scale problem by setting the scale of all output variables to `6.0`.

To strengthen this case we've investigated the optimal scale as determined by the minimal [Jensen-Shannon divergence](https://en.wikipedia.org/wiki/Jensen%E2%80%93Shannon_divergence) and when using `tf.nn.relu` we usually end up with a scale of `8.0`, and with `tf.nn.relu6` we always end up with a scale of `6.0`. This is good since we up with a smaller _ε_, which helps with the precision.

---

Because of the above I am going to put the _DP4A project_ on-hold until we can do a re-write of the neural network code using either newer cuDNN primitives or using a completely custom kernel.

**Comment by Chicoryn** [comment]
After a few busy weeks in San Francisco I am back with some inspiration. I started an experiment with different data types during inference to see the performance difference again, and these are my results so far when using a tiny neural network (no residual blocks to make debugging easier).

The main contribution of this attempt is that the implementation actually works (it predicts the same moves as the float and pseudo-half version), and this network can therefore be used as a valid performance comparison. The technical difference between this and previous implementations are:

- Fixed scaling factor of `6.0` for all output layers (using `tf.nn.relu6` during training), this simplify the code dealing with scaling factors significantly.
- I am using `CUDNN_DATA_INT8x4` and `CUDNN_TENSOR_NCHW_VECT_C` for both the input, output, and filter tensors. This is because of a [comment in tensorflow](https://github.com/tensorflow/tensorflow/blob/3e51dbba08910c14f714d2c68f4d4d2916a2b7eb/tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.cc#L120) that suggests that it is the only working configuration.
- The network was trained using half precision in tensorflow, which may have made it more resilient towards rounding errors.

We can observe roughly a 2x speed-up when using DP4A for this small network, this may grow as the network becomes more compute than host bound. The reason there is not much of a difference between the different data types for smaller batch sizes is because the cost of copying from the host (unpinned) to the device dominates the runtime.

### Int8x4
```
test batch_size_001 ... bench:     100,570 ns/iter (+/- 6,931)
test batch_size_002 ... bench:     102,021 ns/iter (+/- 996)
test batch_size_004 ... bench:     109,425 ns/iter (+/- 1,137)
test batch_size_008 ... bench:     137,083 ns/iter (+/- 1,785)
test batch_size_016 ... bench:     198,625 ns/iter (+/- 3,661)
test batch_size_032 ... bench:     323,130 ns/iter (+/- 3,735)
test batch_size_064 ... bench:     523,549 ns/iter (+/- 7,902)
test batch_size_128 ... bench:     879,324 ns/iter (+/- 10,828)
test batch_size_256 ... bench:   1,591,066 ns/iter (+/- 32,915)
```

### Pseudo-Half
```
test batch_size_001 ... bench:     241,942 ns/iter (+/- 2,780)
test batch_size_002 ... bench:     221,768 ns/iter (+/- 1,990)
test batch_size_004 ... bench:     243,368 ns/iter (+/- 2,819)
test batch_size_008 ... bench:     287,454 ns/iter (+/- 2,473)
test batch_size_016 ... bench:     364,674 ns/iter (+/- 26,745)
test batch_size_032 ... bench:     562,866 ns/iter (+/- 4,068)
test batch_size_064 ... bench:     849,575 ns/iter (+/- 5,168)
test batch_size_128 ... bench:   1,333,253 ns/iter (+/- 8,943)
test batch_size_256 ... bench:   2,321,762 ns/iter (+/- 7,660)
```

### Float
```
test batch_size_001 ... bench:     112,893 ns/iter (+/- 1,711)
test batch_size_002 ... bench:     109,769 ns/iter (+/- 806)
test batch_size_004 ... bench:     147,600 ns/iter (+/- 1,473)
test batch_size_008 ... bench:     212,604 ns/iter (+/- 2,262)
test batch_size_016 ... bench:     319,803 ns/iter (+/- 2,940)
test batch_size_032 ... bench:     539,914 ns/iter (+/- 6,299)
test batch_size_064 ... bench:     866,846 ns/iter (+/- 7,521)
test batch_size_128 ... bench:   1,541,471 ns/iter (+/- 16,759)
test batch_size_256 ... bench:   2,921,560 ns/iter (+/- 15,598)
```

**Comment by Chicoryn** [comment]
There seems to be some _issues_ with the this implementation related to the value head. The observed effect is that occasionally a _random move_ will get a high win rate and after that dominate the rest of the search.

It is unclear if this is due to a bug in the quantization of the value head, or if this is an anomaly in the trained network that does not know how to evaluate certain positions. If it is the later then we have most likely not observed it before now because of the huge performance benefits, which makes it viable for moves that are not recommended by the policy to get promoted.

This behaviour also seems erratic since when going back to a position with a _weird_ value, it is not always possible to re-produce this issue. This may be because of the use of random symmetries.

**Comment by Chicoryn** [comment]
After two weeks of debugging, I think we've found the issue that caused this erratic behaviour. It is related to a race condition in the `Slots` structure, where the CUDA calls are asynchronous **and** we return the memory back to the pool at the end of their lifetime (as defined by the scope).

This means that the CUDA function might get executed **after** the scope has ended (since it is asynchronous), at which point the memory has been returned to the pool and may have been reused by another thread.

The solution is obvious, extend the lifetime of all _slots_ for the full duration of the CUDA calls.

---

After implementing the aforementioned fix I cannot re-produce the issue anymore, and the neural network plays consistently. After a 50 game round-robin between `dg-d-128-2-3`, `d-i-128`, and `leela` this is the final score:

```
dg-i-128 v dg-d-128-2-3 (50/50 games)
unknown results: 4 8.00%
board size: 19   komi: 7.5
               wins              black         white       avg cpu
dg-i-128          5 10.00%       2   8.00%     3  12.00%   3808.76
dg-d-128-2-3     41 82.00%       20 80.00%     21 84.00%    395.15
                                 22 44.00%     24 48.00%

dg-i-128 v leela (50/50 games)
unknown results: 1 2.00%
board size: 19   komi: 7.5
           wins              black         white       avg cpu
dg-i-128     38 76.00%       21 84.00%     17 68.00%   2440.50
leela        11 22.00%       7  28.00%     4  16.00%     71.42
                             28 56.00%     21 42.00%

dg-d-128-2-3 v leela (50/50 games)
board size: 19   komi: 7.5
               wins              black         white       avg cpu
dg-d-128-2-3     47 94.00%       24 96.00%     23 92.00%    276.17
leela             3  6.00%       2   8.00%     1   4.00%     71.19
                                 26 52.00%     24 48.00%
```

These results are consistent with training a normal 128 wide ResNet using floats. They points towards a few issues however:

- The neural network is weak at global thinking, and the extra playouts gained from DP4A does not compensate.
- It does not score all games correctly.

However, these issues are better solved in a different thread. For the moment the DP4A implementation is a success.
